wgpu-core/src/indirect_validation/draw.rs

```diff
@@ -0,0 +1,984 @@
+        }
+
+        let max_staging_buffer_size = 1 << 26; // ~67MiB
+
+        let mut staging_buffers = Vec::new();
+
+        let mut current_size = 0;
+        for batch in batches.values_mut() {
+            let data = batch.metadata();
+            let offset = if current_size + data.len() > max_staging_buffer_size {
+                let staging_buffer =
+                    StagingBuffer::new(device, NonZeroU64::new(current_size as u64).unwrap())?;
+                staging_buffers.push(staging_buffer);
+                current_size = data.len();
+                0
+            } else {
+                let offset = current_size;
+                current_size += data.len();
+                offset as u64
+            };
+            batch.staging_buffer_index = staging_buffers.len();
+            batch.staging_buffer_offset = offset;
+        }
+        if current_size != 0 {
+            let staging_buffer =
+                StagingBuffer::new(device, NonZeroU64::new(current_size as u64).unwrap())?;
+            staging_buffers.push(staging_buffer);
+        }
+
+        for batch in batches.values() {
+            let data = batch.metadata();
+            let staging_buffer = &mut staging_buffers[batch.staging_buffer_index];
+            unsafe {
+                staging_buffer.write_with_offset(
+                    data,
+                    0,
+                    batch.staging_buffer_offset as isize,
+                    data.len(),
+                )
+            };
+        }
+
+        let staging_buffers: Vec<_> = staging_buffers
+            .into_iter()
+            .map(|buffer| buffer.flush())
+            .collect();
+
+        let mut current_metadata_entry = None;
+        for batch in batches.values_mut() {
+            let data = batch.metadata();
+            let (metadata_resource_index, metadata_buffer_offset) =
+                resources.get_metadata_subrange(data.len() as u64, &mut current_metadata_entry)?;
+            batch.metadata_resource_index = metadata_resource_index;
+            batch.metadata_buffer_offset = metadata_buffer_offset;
+        }
+
+        let mut buffer_barriers = Vec::new();
+        let mut buffer_index_set = bit_set::BitSet::new();
+
+        for index in batches
+            .values()
+            .map(|batch| batch.staging_buffer_index)
+            .filter(|index| buffer_index_set.insert(*index))
+        {
+            let staging_buffer = &staging_buffers[index];
+
+            buffer_barriers.push(hal::BufferBarrier {
+                buffer: staging_buffer.raw(),
+                usage: hal::StateTransition {
+                    from: wgt::BufferUses::MAP_WRITE,
+                    to: wgt::BufferUses::COPY_SRC,
+                },
+            });
+        }
+        buffer_index_set.clear();
+
+        for index in batches
+            .values()
+            .map(|batch| batch.metadata_resource_index)
+            .filter(|index| buffer_index_set.insert(*index))
```

Comment:
Could we do these as early-continues in the for loop - as we're iterating already, I think it would read more clearly

Reply:
Related: I was worried about those manual `.clear()` calls. I added 2 new data structures to get rid of the possibility of forgetting to do things in the right order.

Let me know what you think!